Support vagrant-env for configuring DB boxes (#35)

PaulNeumann opened the pull request:
This implements the enhancement in #35 for the single-instance database boxes. It allows configuration via either environment variables or (preferably) the `vagrant-env` plugin, using the same approach as the OLCNE box.

For each box:

- Add `.env` configuration file
- Add or modify `.gitignore` file to ignore `.env.local*`
- Modify `Vagrantfile` to use environment variables or the `vagrant-env` plugin for configuration

No configuration options or defaults have been changed. I've tested these files, and they seem to be working as intended.

**The enhancement is not complete.** If you decide to proceed with it, and if this approach is acceptable, then any feedback on these files needs to be addressed and the `README.md` files need to be updated.

Corrections and suggestions are welcome.

Signed-off-by: Paul Neumann <38871902+PaulNeumann@users.noreply.github.com>

gvenzl commented on OracleDatabase/12.1.0.2/Vagrantfile:
```diff
@@ -20,38 +20,91 @@ VAGRANTFILE_API_VERSION = "2"
+  SYSTEM_TIMEZONE = default_s('SYSTEM_TIMEZONE', host_tz)
+
+  # Oracle base directory
+  # "VM_" prefix used to avoid conflict with host environment variable
+  VM_ORACLE_BASE = default_s('VM_ORACLE_BASE', '/opt/oracle')
```
Why do we need the `VM_` prefix for this? Would this overwrite or collide with already used variables on the host?

PaulNeumann replied:
If one does a lot of Oracle database work, then it's very possible that the ORACLE_BASE, ORACLE_HOME and ORACLE_SID environment variables are set on their host. (They are on mine.) Those environment variables are almost certainly not what we want in the VM, especially if the host is Windows. Using the `VM_` prefix should prevent conflicts.

gvenzl replied:
Ok, thank you, that makes sense. I suppose that would bring us back to the question, or similar question, that @Djelibeybi raised. Should we then not prefix more environment variables? For example, `SYSTEM_TIMEZONE` is opaque as to whether it's the host system or the guest system, so it's probably better to prefix that one as well with `VM_`.

Same with the Oracle related environment variables.  It's equally conceivable that a user has `ORACLE_PDB` also set on his local machine, or `ORACLE_CHARACTERSET`.

Would it not be less confusing prefixing all Vagrant related environment variables with `VM_` then?

Note, that does not mean that we should pass them into the install script as such. There is a benefit not polluting the install script with `VM_` as it can be reused for other tasks. We should still apply a mapping of the "host" `VM_` prefixed environment variables to the actual environment variables within the box, i.e:

```
  config.vm.provision "shell", path: "scripts/install.sh", env:
    {
       "SYSTEM_TIMEZONE"     => VM_SYSTEM_TIMEZONE,
       "ORACLE_BASE"         => VM_ORACLE_BASE,
       "ORACLE_HOME"         => VM_ORACLE_HOME,
       "ORACLE_SID"          => VM_ORACLE_SID,
       "ORACLE_PDB"          => VM_ORACLE_PWD
    }
```

gvenzl commented on OracleDatabase/12.1.0.2/Vagrantfile:
```diff
@@ -20,38 +20,91 @@ VAGRANTFILE_API_VERSION = "2"
+  # VM name
+  NAME = default_s('NAME', 'oracle-12102-vagrant')
+
+  # Memory for the VM (in MB, 2048 MB = 2 GB)
+  MEMORY = default_i('MEMORY', 2048)
```
There is the memory for the VM and there is the memory for the database.
We do not set the memory for the database today but just inherit what `dbca` picks by default (which is 40% of the existing memory on the environment).

`MEMORY` is probably too vague, does it mean VM memory or DB memory? We should consider using a more explicit variable, such as `VM_MEMORY` (I do see the irony of my next comment asking why we have the `VM_` prefix for other variables and am happy to hear other suggestions).

PaulNeumann replied:
`VM_MEMORY` makes sense to me. I'll change it.

PaulNeumann replied:
I've changed `MEMORY` to `VM_MEMORY` in all `.env` and `Vagrantfile` files and updated the PR.

gvenzl commented on OracleDatabase/11.2.0.2/.gitignore:
```diff
@@ -0,0 +1 @@
+.env.local*
```
Why is it here `.env.local*` while the file above is just `.env`? Or does GitHub not render this correctly?

PaulNeumann replied:
`.env` contains the default configuration values. It's committed to the repository, so we don't want Git to ignore it. `.env.local` contains the user's custom configuration values, and shouldn't be committed. I don't know the rationale behind the `*` part; I copied that from the OLCNE `.gitignore` file.

gvenzl commented on OracleDatabase/11.2.0.2/Vagrantfile:
```diff
@@ -20,37 +20,66 @@ VAGRANTFILE_API_VERSION = "2"
+  NAME = default_s('NAME', 'oracle11g-xe-vagrant')
+
+  # Memory for the VM (in MB, 2048 MB = 2 GB)
+  MEMORY = default_i('MEMORY', 2048)
```
Seems like we are duplicating the default values, once here and once in the `.env` file. Is that necessary, as this could easily be overlooked by someone and lead to unexpected behavior?

PaulNeumann replied:
Having the default values in the `Vagrantfile` allows it to work as is, without the user installing the `vagrant-env` plugin or setting any environment variables. If no customization is needed, then people can just run `git clone` / `vagrant up`.

gvenzl commented on OracleDatabase/19.3.0/.env:
```diff
@@ -0,0 +1,53 @@
+# Oracle Database 19.3.0 configuration file
```
Is there a way to inherit environment variables from another file?
A lot of these variable values are the same between 12.1.0.2 and 19.3.
Would be great if there is a way to have "global" defaults that can be overwritten per VM when needed.

PaulNeumann replied:
I'm not sure. I'll look into it, but doing it that way might make it more difficult for people to customize the boxes. It seems like people would either need to create two `.env.local` files at different levels, or combine two `.env` files into a single `.env.local` file. There's some benefit in keeping the boxes atomic.

PaulNeumann replied:
After some thought and experimentation, I recommend keeping the configuration parameters for each box separate. Using inheritance would make configuration more difficult.

Users should be able to modify any/all parameters for each box without affecting any other box. They should be able to do that in one place, without marking the Git tree as changed. In addition, they should be able to run multiple boxes (or multiple copies of the same box) simultaneously.

A `.env` file with common parameters would need to be in a separate directory, for example `vagrant-boxes/OracleDatabase/common`. Then the `.env` file for each box would contain only the parameters unique to that box.

If users wanted to modify both common parameters and box-specific parameters, they would have to copy parameters from both the common `.env` file and the box-specific `.env` file into a single `.env.local` file. This is a lot less convenient (and more error-prone) than just copying `.env` to `.env.local`.

A couple of other considerations also favor keeping the configurations separate:

- Many of the common parameters wouldn't apply to the XE boxes, which could be confusing.
- The common parameters may not always be common. For example, it might be a good idea to increase the default VM memory for the 18c and 19c boxes to 3 GB, to avoid memory warnings during database setup.

I understand the appeal of inheritance, but I think it'd be better to keep the boxes self-contained.

gvenzl replied:
That makes sense, thank you! I just wanted to make sure that we aren't missing a simple trick if there was one.

PaulNeumann commented:
I think this PR is now complete. To recap:

- This PR implements #35 for the six single-instance database boxes.
- It allows using environment variables and/or the `vagrant-env` plugin to configure the installation. Using the plugin is recommended, but optional.
- The approach is identical to the one used for the OLCNE box. Much of the code and README text was copied from the OLCNE box.

For each box:

- Added a `.env` file containing default configuration parameters.
- Added (or modified) a `.gitignore` file to ignore `.env.local*` files. People can use a `.env.local` file to configure the box without marking the Git tree as changed.
- Updated the `README.md` file to describe how to configure the box. (Also added a few newlines around headings and lists to suppress [markdownlint](https://github.com/DavidAnson/markdownlint) warnings.)
- Updated the `Vagrantfile` file to load configuration parameters from, in order of precedence:
  - Environment variables (if set)
  - `.env.local` (if present and if the plugin is installed)
  - `.env` (if the plugin is installed)
  - The default parameters in the `Vagrantfile` itself

Except for renaming the `MEMORY` configuration parameter to `VM_MEMORY` in response to PR feedback above, no configuration options or defaults have been changed.

Each of the boxes can be configured independently, so users can run multiple boxes (or multiple copies of the same box) simultaneously, as long as they configure the boxes to use unique listener and EM Express ports.

I think I've addressed the PR feedback to date, but I'd be grateful for any other corrections or suggestions. @AmedeeBulle, @Djelibeybi, if you have time to take a look, I'd welcome your comments.

Signed-off-by: Paul Neumann 38871902+PaulNeumann@users.noreply.github.com

Djelibeybi commented:
My initial comment is that there appear to be a mix of `VM_` prefixed `ORACLE`-related environment variables and some without the prefix. I think the prefix is useful, but for consistency, it's probably better to make all environment variables (whether they'd conflict or not) that start with `ORACLE` to get a `VM_` prefix.

PaulNeumann commented:
@Djelibeybi Thanks! I'm happy to change this, if desired. My only concern is that this would make the parameter names inconsistent with the variable names used in the `install.sh` scripts. Except for `VM_ORACLE_BASE`, `VM_ORACLE_HOME` and `VM_ORACLE_SID`, the parameter names currently match the shell script variable names. Is there value in preserving that?

Djelibeybi commented:
@PaulNeumann if `install.sh` is in this repo, I think it should all be consistent and non-conflicting. But this isn't my product, so @gvenzl gets to have the final say. :)

PaulNeumann commented:
@gvenzl I think this PR is ready again. I've merged in the libvirt provider changes for the OracleDatabase/19.3.0 box from #242.

In a comment above, @Djelibeybi suggested renaming some additional configuration parameters. My intent was to use the "VM_" prefix only for those parameters whose names are likely to conflict with host environment variables. I think the prefix helps them stand out. That said, I'm happy to make any additional changes that you'd like.

PaulNeumann commented:
@gvenzl I think this PR is ready to go again. I've merged in the box URL changes from #262. I've also updated all repository name references in the `README.md` files from "vagrant-boxes" to "vagrant-projects".

gvenzl commented on OracleDatabase/12.1.0.2/Vagrantfile:
```diff
@@ -20,38 +20,91 @@ VAGRANTFILE_API_VERSION = "2"
+  end
+
+  # VM name
+  NAME = default_s('NAME', 'oracle-12102-vagrant')
```
I think we should probably call this one `VM_NAME` (or whatever prefix we decide on) to make it clear that it's the name of the VM itself.

PaulNeumann commented:
@gvenzl Thanks for your feedback! I've updated the configuration parameter names in the `.env`, `README.md` and `Vagrantfile` files to use the `VM_` prefix as requested. Please let me know if you'd like any additional changes.

gvenzl commented:
Looks great, thank you very, very much @PaulNeumann!

gvenzl commented:
I assume the label "WIP" can go as well, @PaulNeumann? :)

PaulNeumann commented:
@gvenzl Yes, I think the WIP label can be removed. As far as I know, all the comments have been addressed and it's ready to go. Thanks for all your help!

gvenzl commented:
> Thanks for all your help!

No, thank **you** for all the efforts!
